### DISCUSSION TITLE
New rule `exhaustive-deps`

### DISCUSSION THREAD

**Discussion by KubaJastrz** [KubaJastrz/eslint-plugin-react-query#2] [New rule]
## Description

All __static__ variables don't need to be picked up by the rule and put in the `queryKey`.

All __dynamic__ variables should be present in the `queryKey` array.

## Examples

Correct code

```js
import { useQuery } from 'react-query';

// basically everything that is defined outside of the Component is __static__
const staticUrl = 'http://example.com';

function Component({ dynamicVar }) {
  // all variables defined inside Component (or from props) are __dynamic__
  useQuery(['data', dynamicVar], () => fetch(dynamicVar));
  useQuery(['data', dynamicVar], () => fetch(staticUrl + dynamicVar));

  // using only static variables is also fine
  useQuery(['data'], () => fetch(staticUrl));
}
```

Incorrect code

```js
import { useQuery } from 'react-query';

const staticUrl = 'http://example.com';

function Component({ dynamicVar }) {
  useQuery(['data'], () => fetch(dynamicVar));
  useQuery(['data'], () => fetch(staticUrl + dynamicVar));
}
```

## Rationale

It's easy to forget about keys

https://github.com/tannerlinsley/react-query/discussions/1490

**Comment by micsco** [comment]
@KubaJastrz I'm keen to help with this, are you active and around - I can look to get some attempts happening for this, this week

**Reply by KubaJastrz** [reply]
Hello @micsco, thanks for the interest!

I didn't have much time to develop the project recently but all contributions are welcome. I'll definitely take a look if you create a PR or ask for any help.

It's my first eslint plugin, so the codebase may not be the best, but I'm sure we can figure something out if you encounter any problems. I think using [TDD](https://en.wikipedia.org/wiki/Test-driven_development) is the best way to start developing a new rule.

**Reply by KubaJastrz** [reply]
@micsco did you make any progress on this? I'm asking, cause I'd like to move this forward in the near future and help, or perhaps take over if you are okay with it.

**Reply by micsco** [reply]
Hey @KubaJastrz unfortunately priorities changed and I didn't even get a chance to get started on it. 

I don't have anything to offer around this outside of starting some of the framework to it, so best to just start yourself. Apologies I couldn't get to it!
